### PR TITLE
 ivy.el (ivy--reset-state): Filter sorted alist

### LINKED ISSUE #1705: :predicate not called on alists with :sort t
Compare:

```el
(let (called)
  (setq unread-command-events '(?\C-m))
  (ivy-read "" '((c) (a) (b))
            :predicate (lambda (_) (setq called t))
            :sort t)
  called)
  ;; => nil
```

with:

```el
(let (called)
  (setq unread-command-events '(?\C-m))
  (ivy-read "" '((c) (a) (b))
            :predicate (lambda (_) (setq called t))
            :sort nil)
  called)
  ;; => t
```

and:

```el
(let (called)
  (setq unread-command-events '(?\C-m))
  (ivy-read "" '(c a b)
            :predicate (lambda (_) (setq called t))
            :sort t)
  called)
  ;; => t
```

### PR DESCRIPTION
**Pull request by basil-conto**
Fixes #1705 

I've also simplified the surrounding code a bit in the subsequent two commits.

### CODE REVIEW

**Comment by abo-abo** [comment]
Could you please resolve the merge conflict?

File: ivy.el [hunk 1]
```diff
@@ -1929,13 +1929,11 @@ This is useful for recursive `ivy-read'."
             (dynamic-collection
              (setq coll (funcall collection ivy-text)))
             ((and (consp collection) (listp (car collection)))
-             (setq collection
-                   (setf (ivy-state-collection ivy-last)
-                         (if (and sort (setq sort-fn (ivy--sort-function caller)))
-                             (progn
-                               (setq sort nil)
-                               (sort (copy-sequence collection) sort-fn))
```
[the remark this answers is not in the document]
Comment (basil-conto) [reply ivy.el]: > Please be careful about removing `copy-sequence`.

I was, but clearly not enough. I was expecting `cl-remove-if-not` to allocate a new sequence, but I see now that it only does that on demand, which is annoying.

[the remark this answers is not in the document]
Comment (basil-conto) [reply ivy.el]: Wow, `cl-remove` is more broken than I imagined:

```el
(let* ((seq (list 1 2 3))
           (sub (cl-remove 1 seq)))
      (setcar sub 5)
      seq)
  ;; => (1 5 3)
```

I wonder if this counts as a bug.

[the remark this answers is not in the document]
Comment (basil-conto) [reply ivy.el]: In any case, I've re-added the `copy-sequence`.

**Comment by abo-abo** [comment]
Thanks.

**Pull request closed without merging**